### PR TITLE
Index on task instance

### PR DESCRIPTION
**Pull request by babinos87**
Added index `ti_state_incl_start_date` on `taskinstance` table. This index seems to have great positive effect when querying this table in a setup with many millions such rows. Existing index `ti_state_lkp` is not heavily utilized, as the `start_date` column (not part of the existing index) needs to be selected, hence a sequential scan will be performed in a very large table. Including the `start_date` column makes a use of `INDEX ONLY SCAN`, which improves performance dramatically in such situations.

This change is applicable to PostgreSQL database backends only, as other databases do not support this feature.

---

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Review by potiuk — APPROVED** [review]
It looks good,  I do not see any bad side effects of it. Anyone else wants to chime in?

File: airflow/migrations/versions/0125_2_6_0_add_index_to_task_instance_table.py
Comment (ephraimbuddy) [inline airflow/migrations/versions/0125_2_6_0_add_index_to_task_instance_table.py]:
```suggestion
airflow_version = "2.6.1"
```
This should be moved to 2.6.1? cc @potiuk

File: airflow/models/taskinstance.py
Comment (uranusjr) [inline airflow/models/taskinstance.py:433]: Would be a good idea to add a comment here describing what queries this would concretely improve over existing indexes, since it is not immediately obvious.

File: airflow/migrations/versions/0125_2_6_0_add_index_to_task_instance_table.py
Comment (potiuk) [reply airflow/migrations/versions/0125_2_6_0_add_index_to_task_instance_table.py]: yes

File: airflow/models/taskinstance.py
Comment (babinos87) [reply airflow/models/taskinstance.py:433]: Added comment above the index definition, please let me know if this looks OK @uranusjr

File: airflow/migrations/versions/0125_2_6_0_add_index_to_task_instance_table.py
Comment (babinos87) [reply airflow/migrations/versions/0125_2_6_0_add_index_to_task_instance_table.py]: I have tried to modify the filename and the content of `airflow/migration/versions/0125_2_6_0_add_index_to_task_instance.py`, but the `pre-commit run` command keeps creating the `2_6_0` migration file. Is there a way to make it stop doing this and enforce `2_6_1` in the filename? @ephraimbuddy @potiuk

Comment (potiuk) [reply airflow/migrations/versions/0125_2_6_0_add_index_to_task_instance_table.py]: I think it's because we are in a process of releasing 2.6.0 and we have not updated it yet in main (RC1 has just been cut and we **just** branched off.

Can you update it there please? https://github.com/apache/airflow/blob/main/airflow/__init__.py#L29 And also maybe check if there are other places where 2.6.0 is used.

@ephraimbuddy @pierrejeambrun  -> should we have a step for that somewhere in our process  :)?

**Review by uranusjr — APPROVED** [review]
If the `airflow_version` part is figured out.

File: airflow/migrations/versions/0125_2_6_0_add_index_to_task_instance_table.py
Comment (babinos87) [reply airflow/migrations/versions/0125_2_6_0_add_index_to_task_instance_table.py]: Changing https://github.com/apache/airflow/blob/main/airflow/__init__.py#L29 produces a bunch of other changes by `pre-commit run`. Some changes are propose by it, for example:

```
...
Check that the REVISION_HEADS_MAP is up-to-date........................................Failed
- hook id: check-revision-heads-map
- exit code: 3

Current airflow version is not in the REVISION_HEADS_MAP
Current airflow version: 2.6.1
Please add the version to the REVISION_HEADS_MAP at: /home/rais/Desktop/workspace/airflow/airflow/utils/db.py

Update version to the latest version in the documentation..............................Failed
- hook id: update-version
- files were modified by this hook
...
```

(btw `2.6.0` is not listed in https://github.com/apache/airflow/blob/main/airflow/utils/db.py#L55, I am not sure how this works).

And I have find a few other places where `2.6.0` is hard-coded, for example: https://github.com/apache/airflow/blob/main/dev/validate_version_added_fields_in_config.py#L45 - I am not sure whether I would have to change this too?

Should I modify all those files that are produced? I am not sure if this should be done as part of this PR, or someone else should cut a new release? Perhaps I should wait until this happens? I am not very familiar with this procedure. I don't mind to try, but would need to spam you with questions probably. Please let me know how should I proceed.
@ephraimbuddy @potiuk

Comment (babinos87) [reply airflow/migrations/versions/0125_2_6_0_add_index_to_task_instance_table.py]: Hi @ephraimbuddy @potiuk , I see that `2.6.0` has been released, so now I can change https://github.com/apache/airflow/blob/main/airflow/__init__.py#L29 to `2.6.1`? I see it points to `2.7.0-dev`. Not sure how should I patch this for `2.6.1`, is there some part of the documentation that you are aware of which I can look at?

Comment (potiuk) [reply airflow/migrations/versions/0125_2_6_0_add_index_to_task_instance_table.py]: You should change it to 2.7.0. If we decide to add it for 2.6 we will cherry-pick it and modify it accordingly in 2.6 branch. It's at the discretion of the release manager, whether it will be included or not. Mainly based on a risk assesment and whether it will not be classified as new feature (in case we consider it risky or a feature -  2.7.0 will be the time it will go out).

You have to do it similarly to this dicussion: https://github.com/apache/airflow/pull/30827#pullrequestreview-1408586127 and whichever of those two changes will be merged first, the other will have to rebase and resolve conflicts after the first merge.

Comment (babinos87) [reply airflow/migrations/versions/0125_2_6_0_add_index_to_task_instance_table.py]: Thanks for this, it seems very helpful, will do!

**Comment by ashb** [comment]
When do we do any searching on TI by start_date?

File: airflow/migrations/versions/0125_2_6_0_add_index_to_task_instance_table.py
Comment (babinos87) [reply airflow/migrations/versions/0125_2_6_0_add_index_to_task_instance_table.py]: OK, I think that's all, I have revised the commit above too be for `2.7.0`. @potiuk @ephraimbuddy please have a look and let me know if it is OK now

**Comment by dstandish** [comment]
Hi @babinos87 , I happened to be looking a bit at index utilization and noticed that this one isn't getting used on our cluster.  I saw you shared a query that this benefits.  Is that emitted from our codebase?  Or is it just a custom query that you run?  Thanks

**Comment by babinos87** [comment]
@dstandish this is a custom query, we have some custom code run by the scheduler to find the last successful dag run date, which runs this query and this index helps greatly.

**Comment by dstandish** [comment]
Thanks @babinos87 

In that case, I don't think this belongs in airflow so I will plan to remove it.  You can ensure that the index is there yourself as part of your infra process.

Thanks

**Comment by babinos87** [comment]
@dstandish this would make sense if there was a way to define custom sql schema / commands as part of the installation process. As far as I am aware, there is no such thing, and any manual changes could potentially break on db migration. I think it would be more appropriate to plan some work for this, so anyone could use it to define custom schema without risking of breaking the database between upgrades?

**Comment by potiuk** [comment]
> I think it would be more appropriate to plan some work for this, so anyone could use it to define custom schema without risking of breaking the database between upgrades? In any case, I don't foresee any trouble caused by this extra index, and it would help if it stays - my use case is quite basic so I think that someone else could benefit from this too. 

You are free to propose somethign if you could ask - for example - a PR where you would allow anyone to (for example) plug -in  a script that will be used as a step after migration and a way to add it to be able to add after-migration steps, that might be a good contribution that you could do.

As anyone else in the community you can propose PRs there, and it's a nice and isolated one that you are probably one of the best people to test and have need for it, so it would be nice if you could propose one (or find / sponsor/ encourage someone who could contribute it if you do not feel like you have the right skills)

**Comment by dstandish** [comment]
> @dstandish this would make sense if there was a way to define custom sql schema / commands as part of the installation process. As far as I am aware, there is no such thing, and any manual changes could potentially break on db migration. I think it would be more appropriate to plan some work for this, so anyone could use it to define custom schema without risking of breaking the database between upgrades?
> 
> In any case, I don't foresee any trouble caused by this extra index, and it would help if it stays - my use case is quite basic so I think that someone else could benefit from this too.

I don't think it's likely to break anything.  You can simply name it `my_company_index_blah_blah` and we won't clash with it.  Sure if we drop a column that it uses (unlikely) then you might have a problem.  But I think that's I think a risk that you must accept given that you are doing something custom with the scheduler that requires a special index.

Indexes are not free.  When you have more indexes, that is more storage used, more work for postgres on each update, and more for the query planner to sort through.  It doesn't make sense to force all airflow users in the world to have an index that doesn't help them at all, just so that you can use it.  I think we have like 9 indexes or something on task instance?  That's getting quite high.  We can't only add indexes, we must also try to remove ones that are not useful.

Here's the other thing.... The indexes we include in airflow, you know, they might not be perfect for everyone's cluster.  It's possible that some clusters might have an unusual usage pattern and maybe as a result they might need to add special index, even without having custom logic or queries.  And even in that case it might not make sense to add such an index to airflow, given that doing so would be a cost for everyone and a benefit perhaps only to one.  Right?

**Pull request merged**
